### PR TITLE
Simplify nz2nz_z2z-class sparse unary broadcast (and fix #18974)

### LINKED ISSUE #18974: Incorrect element type from specialized sparse unary broadcast methods
The `broadcast` methods specialized for unary operations over `SparseMatrixCSC`s assume that the input matrix's element type is appropriate for the returned matrix (e.g. [here](https://github.com/JuliaLang/julia/blob/6c9f5afa89f451bf2262af6f2bdd87b7b7a328d4/base/sparse/sparsematrix.jl#L1436)). This assumption causes, for example,

``` julia
julia> VERSION
v"0.6.0-dev.979"

julia> sin.(speye(Int, 4))
ERROR: InexactError()
 in convert(::Type{Int64}, ::Float64) at ./float.jl:637
 in _broadcast_unary_nz2z_z2z_T(::Base.#sin, ::SparseMatrixCSC{Int64,Int64}, ::Type{Int64}) at ./sparse/sparsematrix.jl:1425
 in broadcast(::Base.#sin, ::SparseMatrixCSC{Int64,Int64}) at ./sparse/sparsematrix.jl:1403
```

Best!

### PR DESCRIPTION
**Pull request by Sacha0**
Following `@stevengj`'s suggestion in #18975, this pull request makes nz2nz_z2z-class sparse unary broadcast leverage existing broadcast machinery rather than reimplement it. Also fixes #18974.

This implementation is slightly slower than the existing implementation due to the view in `Bnzval = broadcast(f, view(A.nzval, 1:nnz(A)))`; removing the view (`Bnzval = broadcast(f, A.nzval)`) eliminates the performance gap. `Bnzval = broadcast(f, A.nzval); resize!(Bnzval, nnz(A))` might be a reasonable alternative. For discussion of performance (and particularly the apparent need for forcing function-type specialization), see #18975. Best!

### CODE REVIEW

**Comment by tkelman** [comment]
is the view necessary? `nzval` is only ever not that length if the input data is invalid, right?

**Comment by Sacha0** [comment]
> is the view necessary?

Not strictly, no. The view preserves the method's existing behavior, but so would `(Bnzval = broadcast(f, A.nzval); resize!(Bnzval, nnz(A)))`. Preserving the method's existing behavior also doesn't seem strictly necessary?

> nzval is only ever not that length if the input data is invalid, right?

`length(A.nzval) > nnz(A)` is valid I believe?

Thanks for reviewing! Best!

**Comment by tkelman** [comment]
You can have extra space in the nzval array, but there may be places that use nnz and `length(nzval)` interchangeably without checking whether they're the same.

**Comment by stevengj** [comment]
The `resize!` approach seems fine to me; in common cases it will be a no-op.

**Comment by Sacha0** [comment]
Switched to the `resize!` approach. Thanks!

File: base/sparse/sparsematrix.jl [hunk 1]
```diff
@@ -1453,26 +1453,17 @@ round{To}(::Type{To}, A::SparseMatrixCSC) = _broadcast_unary_nz2z_z2z_T(round, A
...
 end
-function _broadcast_unary_nz2nz_z2z{Tv}(f::Function, A::SparseMatrixCSC{Tv})
-    _broadcast_unary_nz2nz_z2z_T(f, A, Tv)
-end
 @_enumerate_childmethods(_broadcast_unary_nz2nz_z2z,
     log1p, expm1, abs, abs2, conj)
-broadcast{TTv}(::typeof(abs2), A::SparseMatrixCSC{Complex{TTv}}) = _broadcast_unary_nz2nz_z2z_T(abs2, A, TTv)
-broadcast{TTv}(::typeof(abs), A::SparseMatrixCSC{Complex{TTv}}) = _broadcast_unary_nz2nz_z2z_T(abs, A, TTv)
-broadcast{TTv<:Integer}(::typeof(abs), A::SparseMatrixCSC{Complex{TTv}}) = _broadcast_unary_nz2nz_z2z_T(abs, A, Float64)
-broadcast{TTv<:BigInt}(::typeof(abs), A::SparseMatrixCSC{Complex{TTv}}) = _broadcast_unary_nz2nz_z2z_T(abs, A, BigFloat)
```
Comment (stevengj) [inline base/sparse/sparsematrix.jl]: Did you mean to delete the specialized `broadcast` methods for `abs` and `abs2` on sparse matrices?

(Eventually I think these should go away in favor of a more general mechanism, but that seems like a matter for another PR?)

**Comment by stevengj** [comment]
PS. Can you file a performance issue with a test case illustrating the slowdown if you omit the `<:Function` signature?  That really needs to be fixed...

File: base/sparse/sparsematrix.jl (hunk 1 above)
Comment (Sacha0) [inline base/sparse/sparsematrix.jl]: The deletion was intentional: The (revised) unspecialized `broadcast` method appears to provide the behavior of the specializations for `abs` and `abs2` (per a little testing at the REPL). More than happy to restore those specializations and nix them elsewhere though if you prefer? Thanks for reviewing!

**Comment by Sacha0** [comment]
> Can you file a performance issue with a test case illustrating the slowdown if you omit the <:Function signature? That really needs to be fixed...

Shall do, likely tomorrow morning. Thanks!

**Comment by tkelman** [comment]
dunno if we have benchmarks for this, but may as well @nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19065#issuecomment-256780896) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/182bd840d9237610dd774e5a918f754da133d950/bfd8cad_vs_72725a1/report.md). cc @jrevels

File: base/sparse/sparsematrix.jl (hunk 1 above)
Comment (stevengj) [inline base/sparse/sparsematrix.jl]: I'm confused, was there already a PR that updated the unspecialized broadcast for unary operations on sparse arrays?

Comment (Sacha0) [inline base/sparse/sparsematrix.jl]: Apologies for the confusion. Clarification: The macro call
```julia
 @_enumerate_childmethods(_broadcast_unary_nz2nz_z2z,
     log1p, expm1, abs, abs2, conj)
```
defines methods
```julia
broadcast(::typeof(abs), A::SparseMatrixCSC) = _broadcast_unary_nz2nz_z2z(abs, A)
broadcast(::typeof(abs2), A::SparseMatrixCSC) = _broadcast_unary_nz2nz_z2z(abs2, A)
```
which, with `_broadcast_unary_nz2nz_z2z` using `Bnzval = broadcast(f, A.nzval)` to determine output type on this branch, seem to provide the behavior of the removed methods,
```julia
broadcast{TTv}(::typeof(abs2), A::SparseMatrixCSC{Complex{TTv}}) = _broadcast_unary_nz2nz_z2z_T(abs2, A, TTv)
broadcast{TTv}(::typeof(abs), A::SparseMatrixCSC{Complex{TTv}}) = _broadcast_unary_nz2nz_z2z_T(abs, A, TTv)
broadcast{TTv<:Integer}(::typeof(abs), A::SparseMatrixCSC{Complex{TTv}}) = broadcast_unary_nz2nz_z2z_T(abs, A, Float64)
broadcast{TTv<:BigInt}(::typeof(abs), A::SparseMatrixCSC{Complex{TTv}}) = broadcast_unary_nz2nz_z2z_T(abs, A, BigFloat)
```
Does that clarify? Best!

Comment (stevengj) [inline base/sparse/sparsematrix.jl]: Ah good.

**Comment by Sacha0** [comment]
The possible performance regressions look like noise.

**Comment by stevengj** [comment]
Yup, they are not things that could have been affected by this.

**Comment by Sacha0** [comment]
Thanks for reviewing / merging!

**Pull request merged**